File: shared/python/cache/redis.py
import logging
from typing import Optional
import redis.asyncio as redis

logger = logging.getLogger(__name__)


class RedisClient:
    def __init__(self, redis_url: str):
        self.redis_url = redis_url
        self.client: Optional[redis.Redis] = None

    async def connect(self) -> None:
        try:
            self.client = redis.from_url(
                self.redis_url,
                encoding="utf8",
                decode_responses=True,
                socket_keepalive=True,
                socket_keepalive_options={1: 1, 2: 3, 3: 3},
            )
            await self.client.ping()
            logger.info("Connected to Redis")
        except Exception as e:
            logger.error(f"Failed to connect to Redis: {e}")
            raise

    async def get(self, key: str) -> Optional[str]:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        return await self.client.get(key)

    async def set(self, key: str, value: str, expiration: int = None) -> None:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        if expiration:
            await self.client.setex(key, expiration, value)
        else:
            await self.client.set(key, value)

    async def delete(self, *keys: str) -> None:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        if keys:
            await self.client.delete(*keys)

    async def exists(self, key: str) -> bool:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        return bool(await self.client.exists(key))

    async def incr(self, key: str) -> int:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        return await self.client.incr(key)

    async def decr(self, key: str) -> int:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        return await self.client.decr(key)

    async def expire(self, key: str, seconds: int) -> None:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        await self.client.expire(key, seconds)

    async def keys(self, pattern: str) -> list[str]:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        return await self.client.keys(pattern)

    async def flush(self) -> None:
        if not self.client:
            raise RuntimeError("Not connected to Redis")
        await self.client.flushdb()

    async def health_check(self) -> bool:
        try:
            if not self.client:
                return False
            await self.client.ping()
            return True
        except Exception as e:
            logger.error(f"Health check failed: {e}")
            return False

    async def close(self) -> None:
        if self.client:
            await self.client.close()
            logger.info("Closed Redis connection")


def cache_key_pattern(service: str, key: str) -> str:
    return f"cache:{service}:{key}"

File: shared/go/cache/redis.go
package cache

import (
	"context"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

type Client interface {
	Get(ctx context.Context, key string) (string, error)
	Set(ctx context.Context, key string, value string, expiration time.Duration) error
	Del(ctx context.Context, keys ...string) error
	Exists(ctx context.Context, key string) (bool, error)
	Incr(ctx context.Context, key string) (int64, error)
	Decr(ctx context.Context, key string) (int64, error)
	Expire(ctx context.Context, key string, expiration time.Duration) error
	Keys(ctx context.Context, pattern string) ([]string, error)
	Flush(ctx context.Context) error
	Close() error
	HealthCheck(ctx context.Context) error
}

type RedisClient struct {
	client *redis.Client
}

func NewRedisClient(redisURL string) (*RedisClient, error) {
	opt, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, fmt.Errorf("failed to parse redis url: %w", err)
	}

	opt.MaxRetries = 3
	opt.PoolSize = 20
	opt.ConnMaxIdleTime = 5 * time.Minute

	client := redis.NewClient(opt)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := client.Ping(ctx).Err(); err != nil {
		return nil, fmt.Errorf("failed to connect to redis: %w", err)
	}

	return &RedisClient{client: client}, nil
}

func (rc *RedisClient) Get(ctx context.Context, key string) (string, error) {
	val, err := rc.client.Get(ctx, key).Result()
	if err == redis.Nil {
		return "", nil
	}
	return val, err
}

func (rc *RedisClient) Set(ctx context.Context, key string, value string, expiration time.Duration) error {
	return rc.client.Set(ctx, key, value, expiration).Err()
}

func (rc *RedisClient) Del(ctx context.Context, keys ...string) error {
	if len(keys) == 0 {
		return nil
	}
	return rc.client.Del(ctx, keys...).Err()
}

func (rc *RedisClient) Exists(ctx context.Context, key string) (bool, error) {
	val, err := rc.client.Exists(ctx, key).Result()
	return val > 0, err
}

func (rc *RedisClient) Incr(ctx context.Context, key string) (int64, error) {
	return rc.client.Incr(ctx, key).Result()
}

func (rc *RedisClient) Decr(ctx context.Context, key string) (int64, error) {
	return rc.client.Decr(ctx, key).Result()
}

func (rc *RedisClient) Expire(ctx context.Context, key string, expiration time.Duration) error {
	return rc.client.Expire(ctx, key, expiration).Err()
}

func (rc *RedisClient) Keys(ctx context.Context, pattern string) ([]string, error) {
	return rc.client.Keys(ctx, pattern).Result()
}

func (rc *RedisClient) Flush(ctx context.Context) error {
	return rc.client.FlushDB(ctx).Err()
}

func (rc *RedisClient) Close() error {
	return rc.client.Close()
}

func (rc *RedisClient) HealthCheck(ctx context.Context) error {
	ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
	defer cancel()

	return rc.client.Ping(ctx).Err()
}

func CacheKeyPattern(service string, key string) string {
	return fmt.Sprintf("cache:%s:%s", service, key)
}
